### PR TITLE
XIVY-10849 proxy based engine probe & download

### PR DESCRIPTION
**Pull request by ivy-rew**

### CODE REVIEW

**Review by ivy-lli — APPROVED** [review]

File: src/main/java/ch/ivyteam/ivy/maven/engine/download/URLEngineDownloader.java [hunk 1]
```diff
@@ -16,60 +16,77 @@
...
-    try (InputStream pageStream = new UrlRedirectionResolver().followRedirections(engineListPageUrl)) {
-      return findEngineDownloadUrl(pageStream);
+    try {
+      var repo = new Repository("engine.list.page", engineListPageUrl.toExternalForm());
+      Path index = Files.createTempFile("page", ".html");
+      wagonDownload(repo, "", index);
+      try (InputStream pageStream = Files.newInputStream(index)) {
+        return findEngineDownloadUrl(pageStream);
+      } finally {
+        Files.delete(index);
```
Comment (alexsuter) [inline src/main/java/ch/ivyteam/ivy/maven/engine/download/URLEngineDownloader.java:67]: check if file exists?

File: src/main/java/ch/ivyteam/ivy/maven/engine/download/URLEngineDownloader.java [hunk 2]
```diff
@@ -80,6 +97,18 @@ private File evaluateTargetFile(URL engineUrl) {
...
 
+  private void wagonDownload(Repository repo, String resource, Path target) throws MojoExecutionException {
+    HttpWagon wagon = new HttpWagon();
+    try {
+      wagon.connect(repo, null, proxies);
+      wagon.get(resource, target.toFile());
+    } catch (Exception ex) {
+      throw new MojoExecutionException("Download failed from repo " + repo + " with resource " + resource, ex);
+    } finally {
+      wagon.closeConnection();
```
Comment (alexsuter) [inline src/main/java/ch/ivyteam/ivy/maven/engine/download/URLEngineDownloader.java:108]: null check? or is new HttpWagon() super safe?

File: src/main/java/ch/ivyteam/ivy/maven/InstallEngineMojo.java [hunk 3]
```diff
@@ -141,6 +142,10 @@ public class InstallEngineMojo extends AbstractEngineMojo {
   @Parameter(property = "ivy.engine.auto.install", defaultValue = "true")
   boolean autoInstallEngine;
 
+  @Component
+  @SuppressWarnings("deprecation")
```
Comment (alexsuter) [inline src/main/java/ch/ivyteam/ivy/maven/InstallEngineMojo.java:146]: deprecated? what should we use instead?

**Review by alexsuter — COMMENTED** [review]
not sure if we should use deprecated stuff here?

File: src/main/java/ch/ivyteam/ivy/maven/engine/download/URLEngineDownloader.java [hunk 4]
```diff
@@ -80,6 +97,18 @@ private File evaluateTargetFile(URL engineUrl) {
     return downloadZip;
   }
 
+  private void wagonDownload(Repository repo, String resource, Path target) throws MojoExecutionException {
+    HttpWagon wagon = new HttpWagon();
+    try {
+      wagon.connect(repo, null, proxies);
+      wagon.get(resource, target.toFile());
```
[the remark this answers is not in the document]
Comment (ivy-rew) [reply src/main/java/ch/ivyteam/ivy/maven/engine/download/URLEngineDownloader.java:104]: yes, redirect support is built in.

File: src/main/java/ch/ivyteam/ivy/maven/engine/download/URLEngineDownloader.java (hunk 2 above)
Comment (ivy-rew) [reply src/main/java/ch/ivyteam/ivy/maven/engine/download/URLEngineDownloader.java:108]: can't think of a situation where this is null. Either the construction fails with an exception or it's not null.

**Pull request merged**